### PR TITLE
Determine elf arch by e_machine and bits

### PR DESCRIPTION
**Pull request by Xeonacid**
# Pwntools Pull Request

Thanks for contributing to Pwntools!  Take a moment to look at [`CONTRIBUTING.md`][contributing] to make sure you're familiar with Pwntools development.

In context module, riscv32 and riscv64 need to be assigned arch separately.
In elf module, riscv32 and riscv64 share the same e_machine, EM_RISCV, and we used e_machine to determine arch directly, which does not work for RISC-V (and MIPS).
Change to determining elf arch by e_machine and bits.

Separate 32/64 for riscv and mips.

## Testing

Pull Requests that introduce new code should try to add doctests for that code.  See [`TESTING.md`][testing] for more information.

## Target Branch

Depending on what the PR is for, it needs to target a different branch.

You can always [change the branch][change] after you create the PR if it's against the wrong branch.

| Branch   | Type of PR                                                       |
| -------- | ---------------------------------------------------------------- |
| `dev`    | New features, and enhancements
| `dev`    | Documentation fixes and new tests
| `stable` | Bug fixes that affect the current `stable` branch
| `beta`   | Bug fixes that affect the current `beta` branch, but not `stable`
| `dev`    | Bug fixes for code that has never been released

[contributing]: https://github.com/Gallopsled/pwntools/blob/dev/CONTRIBUTING.md
[testing]: https://github.com/Gallopsled/pwntools/blob/dev/TESTING.md
[change]: https://github.com/blog/2224-change-the-base-branch-of-a-pull-request

## Changelog

After creating your Pull Request, please add and push a commit that updates the changelog for the appropriate branch.  
You can look at the existing changelog for examples of how to do this.

### CODE REVIEW

**Comment by Xeonacid** [comment]
CHANGELOG.md says `This changelog only includes added major features and changes. Bugfixes and minor changes are omitted.`. But CI fails if I don't add changelog. Should I do this?

File: pwnlib/elf/plt.py [hunk 1]
```diff
@@ -64,8 +64,11 @@ def prepare_unicorn_and_context(elf, got, address, data):
         'arm': U.UC_ARCH_ARM,
         'i386': U.UC_ARCH_X86,
         'mips': U.UC_ARCH_MIPS,
+        'mips64': U.UC_ARCH_MIPS,
         # 'powerpc': U.UC_ARCH_PPC, <-- Not actually supported
         'thumb': U.UC_ARCH_ARM,
+        'riscv32': U.UC_ARCH_RISCV,
+        'riscv64': U.UC_ARCH_RISCV
```
Comment (Arusekk) [inline pwnlib/elf/plt.py]: just a nitpick
```suggestion
        'riscv64': U.UC_ARCH_RISCV,
```

**Review by Arusekk — APPROVED** [review]
Great stuff! I really like what you have done here. One minor thing though.

**Comment by Arusekk** [comment]
Please see why the ROP tests suddenly fail (I think this is something about how your changes influence GOT/PLT resolution).

**Comment by Xeonacid** [comment]
> Please see why the ROP tests suddenly fail (I think this is something about how your changes influence GOT/PLT resolution).

Unfortunately, I can't reproduce any of the failures on my machine. Could you try a reproduction please?

**Comment by peace-maker** [comment]
Cool stuff, used it already during hack-a-sat. I have a follow up patch ready which fixes shellcode assembly and debugging.
All other architectures are named without the `32` suffix like `mips`/`mips64`, `powerpc`/`powerpc64`, and `sparc`/`sparc64`. Is it intentional to break this pattern in `riscv32`/`riscv64`?

**Comment by Xeonacid** [comment]
> All other architectures are named without the `32` suffix like `mips`/`mips64`, `powerpc`/`powerpc64`, and `sparc`/`sparc64`. Is it intentional to break this pattern in `riscv32`/`riscv64`?

The name is picked from gcc target triple or `uname -m`. Nobody calls the 32bit variant `riscv` but `riscv32`.
Those arches which don't have a `-32` postfix should be a historical reason when there is no 64bit variant I think. Newer arches like `riscv` and `loongarch` have the postfix.

**Comment by Arusekk** [comment]
Alright, this is definitely a new feature, not a bugfix, so it should go in dev.
RISC-V support in Unicorn appeared in 2.0.0, which does not support py2 because nobody checked it (see https://github.com/unicorn-engine/unicorn/pull/1822).

**Comment by Xeonacid** [comment]
> Alright, this is definitely a new feature, not a bugfix, so it should go in dev.

I this this is bugfix more than a new feature. Current stable fails with a simple riscv64 binary:
```python
>>> context.binary = ELF('rv64')
[!] Could not populate PLT: AttributeError: arch must be one of ['aarch64', 'alpha', 'amd64', 'arm', 'avr', 'cris', 'i386', 'ia64', 'm68k', 'mips', 'mips64', 'msp430', 'none', 'powerpc', 'powerpc64', 'riscv', 's390', 'sparc', 'sparc64', 'thumb', 'vax']
[*] '/tmp/rv64'
    Arch:     em_riscv-64-little
    RELRO:    Partial RELRO
    Stack:    No canary found
    NX:       NX enabled
    PIE:      PIE enabled
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/pwnlib/context/__init__.py", line 785, in arch
    defaults = self.architectures[arch]
KeyError: 'em_riscv'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/site-packages/pwnlib/context/__init__.py", line 176, in fset
    self._tls[name] = validator(self, val)
  File "/usr/lib/python3.10/site-packages/pwnlib/context/__init__.py", line 872, in binary
    self.arch   = binary.arch
  File "/usr/lib/python3.10/site-packages/pwnlib/context/__init__.py", line 176, in fset
    self._tls[name] = validator(self, val)
  File "/usr/lib/python3.10/site-packages/pwnlib/context/__init__.py", line 787, in arch
    raise AttributeError('AttributeError: arch must be one of %r' % sorted(self.architectures))
AttributeError: AttributeError: arch must be one of ['aarch64', 'alpha', 'amd64', 'arm', 'avr', 'cris', 'i386', 'ia64', 'm68k', 'mips', 'mips64', 'msp430', 'none', 'powerpc', 'powerpc64', 'riscv', 's390', 'sparc', 'sparc64', 'thumb', 'vax']
```

**Comment by Arusekk** [comment]
Also, is rv64 ISA really different from rv32? Or should they just be both treated as `riscv` by pwntools code? I think they can have common shellcodes in the future etc, and `bits` is used anyway in all places to distinguish architectures.

**Comment by Xeonacid** [comment]
> I think they can have common shellcodes in the future etc

They may share the same shellcode for some commands, but I don't think all would work.
For example, riscv32 use `lw/sw` (32 bit) for loading/storing, while riscv64 is likely to use `ld`/`sd` (64 bit).

File: pwnlib/context/__init__.py [hunk 2]
```diff
@@ -404,7 +404,8 @@ class ContextType(object):
         'msp430':    little_16,
         'powerpc':   big_32,
         'powerpc64': big_64,
-        'riscv':     little_32,
+        'riscv32':   little_32,
+        'riscv64':   little_64,
```
Comment (Xeonacid) [inline pwnlib/context/__init__.py:408]: > Also, is rv64 ISA really different from rv32? Or should they just be both treated as riscv by pwntools code? I think they can have common shellcodes in the future etc, and bits is used anyway in all places to distinguish architectures.

And what should we do here? If hard-coded `riscv` as 32 or 64, ppl have to set `context.bits` manually to work for the other one.

**Comment by Arusekk** [comment]
Just as they need to do anyway with e.g. mips endianness.  The recommended way to set everything automatically is using context.binary anyway.

Or we can add aliases rv32/rv64/riscv32/riscv64 to set both architecture and bits.

**Comment by Xeonacid** [comment]
> Just as they need to do anyway with e.g. mips endianness.  The recommended way to set everything automatically is using context.binary anyway.
> 
> Or we can add aliases rv32/rv64/riscv32/riscv64 to set both architecture and bits.

Alright. Then could you introduce a right approach for adding aliases? Seems to need some refactoring.

**Pull request merged**